README.md
# Progress Auditor
*CSE 197 Submission Management Program*

Created by Professor Mia Minnes

Developed by Andrew Kuo

---

## Purpose
To make a program that streamlines and tracks the grading and logistics of discussion-based turnin processes.

## Table of Contents
* [Usage](#usage)
* [Getting Started](#setting-up-the-workspace)
* [Running the Program](#running-the-program)
* [Building for Production](#building-for-production)
* [Project Structure](#project-structure)
  * [Entry Point](#entry-point)

---

## Compatibility
- [x] Windows
- [x] Mac
- [x] Linux

---

## Usage

You will need 2 things:
- The executable for your platform
- A config file called 'config.json' next to it

Make sure to choose the correct executable for your platform. You can double-click to run them in interactive mode or execute them through the terminal.

There are generally 3 types of prompts:
- **Confirmation prompts**: answer with 'y' or 'n' (no need to press 'enter').
- **Select prompts**: choose with 'up' or 'down' arrow keys. Press 'enter' to select. If multiple answers are allowed, use 'space' to select multiple choices and press 'enter' when done.
- **Input promps**: type in your answer and press 'enter' when done.

The general rule is to:
- Press 'enter' to confirm your answer
- Use 'up' and 'down' to navigate
- If multiple answers are allowed, use 'space' to select then press 'enter' to finalize.

**NOTE**: If you want to cancel the operation at anytime, press 'escape'.

Here are the available options for the config file:

| Name              | Description  | Examples |
| ----------------- | ------------ | -------- |
| scheme            | The review procedures to execute. Currently, only 'piazza' is available. | "piazza" |
| inputPath         | The path to the input files. | ./inputs/ |
| outputPath        | The path to the output files. | ./outputs/ |
| currentDate       | The date of the execution. Could be used to simulate older executions. | "2019-01-01" |
| includes          | An array of file paths to other configs. This will load and merge the configs in these directories. The top-level, the current config, will override all included configs and so on. | [ "./base", "../siblingPath/otherDirectory" ] |
| assignments       | An array of assignment entry objects. | /* Refer below. */ |
| inputs            | An array of input entry objects. | /* Refer below. */ |
| outputs           | An array of output entry objects. | /* Refer below. */ |
| outputAutoDate    | Whether to write output into a directory named by date of execution. | true |
| debug             | Whether to enable debug mode for more debugging information, such as stack traces. To output debug logs, add a debug entry to the outputs instead. | false |

Here's an example config file:
```json
{
    "currentDate": "2020-01-01",
    "include": [],
    "scheme": "piazza",
    "inputPath": "./inputs",
    "outputPath": "./outputs",
    "outputAutoDate": true,
    "debug": false,
    "assignments": [
        {
            "assignmentName": "intro",
            "pattern": "intro",
            "opts": {}
        },
        {
            "assignmentName": "week",
            "pattern": "sunday",
            "opts": {}
        },
        {
            "assignmentName": "last",
            "pattern": "last",
            "opts": {}
        }
    ],
    "inputs": [
        {
            "inputName": "reviews.csv",
            "parser": "reviews",
            "opts": {}
        },
        {
            "inputName": "contributions.csv",
            "parser": "contributions",
            "opts": {}
        },
        {
            "inputName": "cohort.csv",
            "parser": "cohort",
            "opts": {
                "maxEndDates": [
                    { "pattern": "Special Summer Session", "endDate": "2020-10-10" }
                ]
            }
        }
    ],
    "outputs": [
        {
            "outputName": "debug",
            "format": "debug"
        },
        {
            "outputName": "slip-days.csv",
            "format": "instructor"
        },
        {
            "outputName": "reports.csv",
            "format": "student",
            "opts": {
                "exportPDF": "reports.pdf",
                "customIntro": "Good morning!",
                "customOutro": "Have a nice day!"
            }
        }
    ]
}
```

### Assignment Entries
| Property  | Description |
| --------- | ----------- |
| assignmentName | The unique assignment name. |
| pattern    | The due date assigning pattern type. The currently implemented parsers are `intro`, `sunday`, and `last`. |
| opts      | Any additional options for the assignment. |
| opts.customPath | You can specify a custom script by setting this to the path to the custom assigner. |

Example:
```json
{
    "assignmentName": "week",
    "pattern": "sunday",
    "opts": {}
}
```

### Input Entries
| Property  | Description |
| --------- | ----------- |
| inputName | The input file name or path relative to the `inputPath`. |
| parser    | The parser type. The currently implemented parsers are `cohort`, `contributions`, and `reviews`. |
| opts      | Any additional options for the input file. |
| opts.customPath | You can specify a custom script by setting this to the path to the custom parser. |
| opts.maxEndDates | You can specify an array of maximum end dates for any user given the matching pattern. Each entry takes a regex pattern to match and a date. This is only implemented for cohort. |

Example:
```json
{
    "inputName": "cohort.csv",
    "parser": "cohort",
    "opts": {}
}
```

#### Max End Date Entries
These entries are only available for the cohort parser.

| Property  | Description |
| --------- | ----------- |
| pattern | The regex pattern to match in cohort for each user if they should have their schedule be limited by the max end date. |
| endDate | The maximum possible end date for the matched users. |

Example:
```json
{
    "inputName": "cohort.csv",
    "parser": "cohort",
    "opts": {
        "maxEndDates": [
            { "pattern": "Special Summer Session 2018", "endDate": "2018-7-11" },
            { "pattern": "Special Summer Session 2019", "endDate": "2019-8-23" },
            { "pattern": "Winter 2020", "endDate": "2020-03-15" }
        ]
    }
}
```

### Output Entries
| Property  | Description |
| --------- | ----------- |
| outputName | The output file name or path relative to the `inputPath`. |
| format    | The format type. The currently implemented parsers are `instructor`, `student`, and `debug`. |
| opts      | Any additional options for the output file. |
| opts.customPath | You can specify a custom script by setting this to the path to the custom output. |
| opts.exportPDF | The PDF output file name for the `student` output type. This will enable PDF export. |
| opts.customIntro | A custom intro to prepend to each PDF output. |
| opts.customOutro | A custom intro to append to each PDF output. |

Example:
```json
{
    "outputName": "reports.csv",
    "format": "student",
    "opts": {
        "exportPDF": "reports.pdf",
        "customIntro": "Hello!",
        "customOutro": "Goodbye!"
    }
}
```

## Contributing

### Installing [Node.js](https://nodejs.org/en/)
This is required to test the program. Just get the current version and install it.

### Installing [Git](https://git-scm.com/)
This is required to edit the program remotely. Just get the current version and install it. The repository is hosted at [GitHub](https://github.com/andykuo1/progress-auditor).

### Installing [VS Code](https://code.visualstudio.com/)
This is not required, but recommended (by me). Just get the current version and install it.

Otherwise, you just need a text editor to write JavaScript, HTML, and CSS.

**Note:** Be sure to get the compatible versions for your operating system.

#### Recommended VS Code Packages
* [Excel Viewer](https://marketplace.visualstudio.com/items?itemName=GrapeCity.gc-excelviewer)
  * Let's you look at CSV files with filters.
* [vscode-pdf](https://marketplace.visualstudio.com/items?itemName=tomoki1207.pdf)
  * Allows you to open and view PDF files.

### Getting the remote repository
Open a command line or terminal and enter a directory to where to copy the project repository. This can be anywhere in your local file system (like your home directory). For example:

```
cd ~/
```

Then, clone the [repo](https://github.com/andykuo1/progress-auditor.git) to the directory.

```
git clone https://github.com/andykuo1/progress-auditor.git
```

Navigate into the directory of the repository.

```
cd progress-auditor
```

To ensure and verify the state of the repository enter the following command:

```
git status
```

### Installing dependencies
Open a command line or terminal and enter into the project directory. This should be where you've copied the remote repository. Following the previous example:

```
cd ~/progress-auditor
```

If you want to inspect the contents of this directory, it should contain the project files, such as `package.json`.

Then run the following command:

```
npm install
```

This should automatically start installing the dependencies (as listed in `package.json`). After it finishes, it should create a directory called `node_modules`, which contains all required dependencies.

**Note:** The `node_modules` directory sometimes contains files unique to each platform so this directory SHOULD NOT be committed to the repository.

**Note:** If a `package-lock.json` is created, it should be committed to the repo. It should not be ignored.

After that, the project is ready to run. _Happy coding!_

---

## Running the Program
After saving any changes to a file, open a command line or terminal and enter into the project directory.

**Note:** If using the recommended `VS Code` package, the in-editor terminal is automatically opened at the project directory. (No need to `cd` every time!)

### Production Build
To "compile" the scripts for public distribution:

```
npm run build
```

**Note:** This will bundle all the resources and assets required into `dist`. It will also "uglify" the code to reduce size and apply other optimizations.

Then, open the program. Either by just opening the outputted executable file itself or running the command:

```
npm start
```

### Development Build
Currently there is no other way to build the project. Due to its rather small size, building the production build every time is not a huge concern.

---

## Building for Production
First, bump the version number in `package.json` and update the `CHANGELOG.md` with the new additions, removals, and fixes.

Then, run this to make sure all dependencies are installed correctly (or update them if necessary). 

```
npm install
```

Then, build and make sure it compiles:

```
npm run start
```

Then, compile build files for production.

```
npm run build-all
```

This will generate the files in `dist` for each supported platform. These files should be not be pushed, but keep these files for later.

You should now be able to commit and push your these changes.

Once done, we will upload the output files generated earlier as a release (be sure the version is correct).

And that should be all :)

---

## Project Structure

### Entry point
The entry point for the code is in `src/bin.js` (if bundled, this will be referred to by `package.json` through the property `bin`).

This script manages the interactive menu and handles starting the program. For the actual implementations, you should checkout `main.js` and `index.js`. `index.js` handles the imports/exports used and `main.js` handles the lifecycle of the program itself.

### Conclusion
If you have any more questions, please contact Professor Mia Minnes.

Or, you can contact me: ank060@ucsd.edu

_(Please begin the subject with 'UCSD Progress Auditor')_

Thank you for reading me!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.2] - 2020-06-09
### Fixed
- Fixed undefined access of "type" for review selection due to type mismatch from enquirer

## [1.0.1] - 2020-06-07
### Added
- Users can now restart on error instead of force exit.

### Fixed
- Missing input files should now properly ask to skip instead of error out.
- Show proper error message for missing parser type.

## [1.0.0] - 2019-09-24
### Fixed
- Auto-assign for last week or week[n + 1] now assigns properly.

### Changed
- Released version 1.0.0
- And more documentation throughout.
- `ignore_review` now keeps old reviews around for posterity instead of removing them.

## [0.4.3] - 2019-09-22
### Changed
- Batch processing now applies to all selected automatically
- Updated index exports to use new utilities

### Fixed
- Removed false-negative error for add_owner_key.

### Removed
- `change_assignment` no longer has history tracking (but applies to all automatically instead)

## [0.4.2] - 2019-09-17
### Added
- `change_assignment` review now has history tracking

### Fixed
- Missing version number in the title
- Post number propagation now works correctly for all assignment types
- PDF exporting will now check for overwrite as well

## [0.4.1] - 2019-09-17
### Added
- Another save-review check if review session fails
- Added more auto-complete for other review types
- Stack traces for unexpected errors

### Changed
- User ID is now PID instead of e-mail

### Fixed
- Missing error count output for review session

## [0.4.0] - 2019-09-16
### Added
- Now has a safety net to log all data when it crashes
- New testing environment
- New config resolution technique when things are missing
- New review maker that can be customized for every review
- New UI system
- New interactive review build system
- New review form completion
- New base scheme that handles generic reviews
- Allow empty reviews (for type 'empty')
- Error skipping
- New review for ignoring other reviews (reflection-ish)
- The Progress Otter wants to say more things
- Escape at any time
- Autocomplete system for options
- Batch error processing by type
- Cache output on error
- Force check for overwriting files

### Changed
- A decent refactor of handlers and loaders
- Vacations now live in reviews for convenience
- Vacation padding is now applied at the top level config
- Move review saves to before error logs

### Removed
- All resolvers are now reviews without builders
- Scheme handler, use Review Registry instead (they are still called schemes though)

### Fixed
- Assignment changes for "last" now propagate correctly

## [0.3.2] - 2019-09-09
### Added
- Allow absolute paths for config paths
- maxEndDates option for cohort parsing

### Changed
- Auto date for output no longer uses time, only dates
- Review outputs now save the ENTIRE review database to file

### Fixed
- Weird month offset for assignments

## [0.3.1] - 2019-09-09
### Fixed
- Executable executes at (seemingly) random directory
- Executable crashes for exporting PDF

## [0.3.0] - 2019-09-08
### Added
- Output can now auto-append dates
- Inputs and outputs can have custom defined scripts
- Default implementations for assignment patterns
- Default implementations for input formats
- PDF exporting for student reports
- Added stack tracing for debug mode

### Changed
- Revamped API system (again with less clutter)
- Selection lists have END dividers
- Debug output now always confirms with user before saving
- Allow input file paths to not exist and be ignored instead
- Added ProgressOtter to the battlefront
- Moved more info to before the error review
- Added submission date change review type
- Disabled default config

### Fixed
- Duplicate reviews should not crash
- Flipped review signature documentation
- Reworded "Ill-formatted" to make more sense

### Removed
- Null review type from the review selection list

## [0.2.0] - 2019-08-25
### Added
- CLI Error handling
- CLI Review system
- Scheme system (with default Piazza scheme)
- Add Submission review type
- Change Assignment Status review type
- Output now shows name of user
- Output now shows how many assignments are due

### Changed
- Revamped API system (hopfully less complicated)
- Due date to count from UTC-12 to compensate for all time zones
- Reoreded output to most recent due -> first submission
- Made all databases to use owner keys when possible
- Average slip day calculations are now done as medians

### Fixed
- Fix output not creating parent directories as well

### Removed
- No longer support reviewer and resolver external scripts

## [0.1.0] - 2019-08-11
### Added
- Generate instructor report for users
- Evaluate vacations for users
- Generate schedules based on user information
- Generic review process system
- Process submissions for assignments

[Unreleased]: https://github.com/andykuo1/progress-auditor/compare/v1.0.1...HEAD
[1.0.1]: https://github.com/andykuo1/progress-auditor/releases/tag/v1.0.1
[1.0.0]: https://github.com/andykuo1/progress-auditor/releases/tag/v1.0.0
[0.4.3]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.4.3
[0.4.2]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.4.2
[0.4.1]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.4.1
[0.4.0]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.4.0
[0.3.2]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.3.2
[0.3.1]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.3.1
[0.3.0]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.3.0
[0.2.0]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.2.0
[0.1.0]: https://github.com/andykuo1/progress-auditor/releases/tag/v0.1.0

src/test1.js
const { printTable, printDivider, printRow } = require('./output/console.js');
const { createSchedule, calculateNumberOfSlipDays } = require('./schedule.js');
const { generateAssignments } = require('./assignment.js');

const WEEK_DAYS = ['U', 'M', 'T', 'W', 'R', 'F', 'S'];
function dateString(date)
{
    return (date.getMonth() + 1) + '/' + date.getDate() + ':' + WEEK_DAYS[date.getDay()];
}

function generateUserReport(db, userID)
{
    const userData = ['Andrew', new Date('Jul 01, 2019'), new Date('Aug 01, 2019')];
    const schedule = createSchedule(userData[1], userData[2], { threshold: 2 });
    const assignments = generateAssignments(schedule);

    {
        const LENGTH = 20;
        printDivider(LENGTH);
        printRow('    Name: ' + userData[0], LENGTH);
        printRow('   Start: ' + dateString(userData[1]), LENGTH);
        printRow('     End: ' + dateString(userData[2]), LENGTH);
        printRow('MaxSlips: ' + calculateNumberOfSlipDays(schedule), LENGTH)
        printDivider(LENGTH);
        console.log();
    }

    generateUserSchedule(db, schedule, assignments);
    console.log();
    generateSubmissionReport(db, userData, schedule, assignments);
    console.log();
}

function generateSubmissionReport(db, userData, schedule, assignments)
{
    const submissions = new Map();
    submissions.set('intro', {
        status: 'Y',
        postponed: 0,
        timestamp: new Date('July 06, 2019')
    });
    submissions.set('week1', {
        status: 'Y',
        postponed: 0,
        timestamp: new Date('July 14, 2019')
    });
    submissions.set('week2', {
        status: 'Y',
        postponed: 0,
        timestamp: new Date('July 21, 2019')
    });
    submissions.set('week3', {
        status: 'Y',
        postponed: 0,
        timestamp: new Date('July 28, 2019')
    });
    submissions.set('week4', {
        status: 'Y',
        postponed: 0,
        timestamp: new Date('Aug 04, 2019')
    });
    submissions.set('week5', {
        status: 'Y',
        postponed: 0,
        timestamp: new Date('Aug 11, 2019')
    });
    submissions.set('last', {
        status: 'N',
        postponed: 0,
        timestamp: new Date('Aug 18, 2019')
    });

    // Compute statuses for each submission...
    {
        const header = ['Assign', 'Due', 'Submit', 'Status', 'Postponed'];
        const entries = [];
        for(const assignment of assignments)
        {
            const assignmentSubs = submissions.get(assignment.name);
            const submission = Array.isArray(assignmentSubs) ? assignmentSubs[0] : assignmentSubs;
            entries.push([assignment.name, dateString(assignment.date), dateString(submission.timestamp), submission.status]);
        }
        printTable(entries, header);
    }
}

function generateUserSchedule(db, schedule, assignments)
{
    const dueDates = [];
    const header = [];
    for(const assignment of assignments)
    {
        header.push(assignment.name);
        dueDates.push(dateString(assignment.date));
    }
    printTable([dueDates], header);
}

generateUserReport();


// COMPLETE = 0x2713 (checkmark)
const COMPLETE_TOKEN = '\u2713';
// INCOMPLETE = 0x2717 (cross) (RED)
const INCOMPLETE_TOKEN = '\u2717';
// UNASSIGNED = _ (empty)
const UNASSIGNED_TOKEN = '\u25A0';
// INREVIEW = ?
const INREVIEW_TOKEN = '?';
// POSTPONED = ...
const POSTPONED_TOKEN = '...';
// OUTOFBOUNDS = 0x25A0 (filled square) (DARK)
const OUTOFBOUNDS_TOKEN = '_';

function generateWeekStatusReport(userID, maxWeeks)
{

}

function generateInstructorReport(db)
{
    // Show Key
    const OUTPUT_LENGTH = 24;
    printDivider(OUTPUT_LENGTH);
    printRow('Key', OUTPUT_LENGTH);
    printDivider(OUTPUT_LENGTH);
    printRow(`${COMPLETE_TOKEN}: Complete`, OUTPUT_LENGTH);
    printRow(`${INCOMPLETE_TOKEN}: Incomplete`, OUTPUT_LENGTH);
    printRow(`${UNASSIGNED_TOKEN}: Unassigned`, OUTPUT_LENGTH);
    printRow(`${INREVIEW_TOKEN}: In-review`, OUTPUT_LENGTH);
    printRow(`${OUTOFBOUNDS_TOKEN}: Out-of-bounds`, OUTPUT_LENGTH);
    printRow(`${POSTPONED_TOKEN}: Vacation`, OUTPUT_LENGTH);
    printDivider(OUTPUT_LENGTH);
    console.log();

    // Show table
    const table = [
        ['Andrew', '3/4', COMPLETE_TOKEN, INCOMPLETE_TOKEN + '(-1)', UNASSIGNED_TOKEN, UNASSIGNED_TOKEN, OUTOFBOUNDS_TOKEN, OUTOFBOUNDS_TOKEN],
        ['Cookies', '2/4', INCOMPLETE_TOKEN + '(-2)', UNASSIGNED_TOKEN, UNASSIGNED_TOKEN, OUTOFBOUNDS_TOKEN, OUTOFBOUNDS_TOKEN, OUTOFBOUNDS_TOKEN],
        ['Pancakes', '3/4', COMPLETE_TOKEN + '(-1)', COMPLETE_TOKEN, UNASSIGNED_TOKEN, OUTOFBOUNDS_TOKEN, OUTOFBOUNDS_TOKEN, OUTOFBOUNDS_TOKEN],
        ['Waffles', '-3/4', COMPLETE_TOKEN + '(-3)', COMPLETE_TOKEN + '(-4)' + INREVIEW_TOKEN, POSTPONED_TOKEN, UNASSIGNED_TOKEN, UNASSIGNED_TOKEN, OUTOFBOUNDS_TOKEN],
    ];

    const COL_HEADERS = ['Name', 'Slips', 'W0', 'W1', 'W2', 'W3', 'W4', 'W5'];

    printTable(table, COL_HEADERS);
    console.log();
}

// generateInstructorReport();

/*
console.log("Example schedule for...");
const startDate = new Date('2019-07-03 06:00:01 UTC');
const endDate = new Date('2019-09-23 00:00:01 UTC');
const dueDates = generateDueDates(startDate, endDate);
console.log("START:", startDate);
console.log("END:", endDate);
console.log(dueDates.join('\n'));
*/
